packages/flutter/lib/src/rendering/object.dart

```diff
@@ -2701,10 +2701,31 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   ///
   /// Used by coordinate conversion functions to translate coordinates local to
   /// one render object into coordinates local to another render object.
+  ///
+  /// Some RenderObjects will provide a zeroed out matrix in this method,
+  /// indicating that even though
   void applyPaintTransform(covariant RenderObject child, Matrix4 transform) {
     assert(child.parent == this);
   }
 
+  /// Whether the given child would be painted if [paint] were called.
+  ///
+  /// Some RenderObjects skip painting their children if they are configured to
+  /// not produce any visible effects. For example, a [RenderOffstage] with
+  /// its `offstage` property set to true, or a [RenderOpacity] with its opacity
+  /// value set to zero.
+  ///
+  /// In these cases, the parent may still supply a non-zero matrix in
+  /// [applyPaintTransform] to inform callers about where it would paint the
+  /// child if the child were painted at all. Alternatively, the parent may
+  /// supply a zeroed out matrix if it would not otherwise be able to determine
+  /// a valid matrix for the child and thus cannot meaningfully determine where
+  /// the child would paint.
+  bool paintsChild(covariant RenderObject child) {
```

Comment:
is there an easy way to check whether the layout will be offscreen? or if there is a parent with clip that clipped the whole renderobject? it may be a good default implementation

Reply:
No.

RenderOffstage and RenderOpacity, for example, return the same exact transform/clip whether they are going to paint their children or not.

Offstage does not really mean the same as offscreen - it's not as if you're getting laid out or painted in some coordinates that are outside of the current viewport. It's just that you will never get `paint` called after being laid out.